systems/controllers/osc/osc_tracking_data_mbp.h

```diff
@@ -0,0 +1,377 @@
+  // void UpdateWeight(Eigen::MatrixXd W);
+
+  // Getters
+  std::string GetName() { return name_; };
+  int GetTrajDim() { return n_r_; };
+  bool IsActive() { return track_at_current_state_; }
+
+  void SaveDdyCommandSol(const Eigen::VectorXd& dv);
+
+  // Print feedback and desired values
+  void PrintFeedbackAndDesiredValues(const Eigen::VectorXd& dv);
+
+  // Finalize and ensure that users construct OscTrackingDataMBP class
+  // correctly.
+  void CheckOscTrackingData();
+
+ protected:
+  int GetStateIdx() { return state_idx_; };
+  void AddState(int state);
+
+  // Feedback output, Jacobian and dJ/dt * v
+  Eigen::VectorXd error_y_;
+  Eigen::VectorXd error_dy_;
+  Eigen::VectorXd y_;
+  Eigen::VectorXd dy_;
+  Eigen::MatrixXd J_;
+  Eigen::VectorXd JdotV_;
+
+  // Desired output
+  Eigen::VectorXd y_des_;
+  Eigen::VectorXd dy_des_;
+  Eigen::VectorXd ddy_des_;
+  Eigen::VectorXd ddy_des_converted_;
+
+  // Commanded acceleration after feedback terms
+  Eigen::VectorXd ddy_command_;
+  // Osc solution
+  Eigen::VectorXd ddy_command_sol_;
+
+  // `state_` is the finite state machine state when the tracking is enabled
+  // If `state_` is empty, then the tracking is always on.
+  std::vector<int> state_;
+
+  /// OSC calculates feedback positions/velocities from `tree_w_spr_`,
+  /// but in the optimization it uses `tree_wo_spr_`. The reason of using
+  /// RigidBodyTree without spring is that the OSC cannot track desired
+  /// acceleration instantaneously when springs exist. (relative degrees of 4)
+  const drake::multibody::MultibodyPlant<double>* plant_w_spr_;
+  const drake::multibody::MultibodyPlant<double>* plant_wo_spr_;
+
+  // World frames
+  const drake::multibody::BodyFrame<double>& world_w_spr_;
+  const drake::multibody::BodyFrame<double>& world_wo_spr_;
+
+ private:
+  // Check if we should do tracking in the current state
+  void UpdateTrackingFlag(int finite_state_machine_state);
+
+  // Updaters of feedback output, jacobian and dJ/dt * v
+  virtual void UpdateYAndError(
+      const Eigen::VectorXd& x_w_spr,
+      drake::systems::Context<double>& context_w_spr) = 0;
+  virtual void UpdateYdotAndError(
+      const Eigen::VectorXd& x_w_spr,
+      drake::systems::Context<double>& context_w_spr) = 0;
+  virtual void UpdateYddotDes() = 0;
+  virtual void UpdateJ(const Eigen::VectorXd& x_wo_spr,
+                       drake::systems::Context<double>& context_wo_spr) = 0;
+  virtual void UpdateJdotV(const Eigen::VectorXd& x_wo_spr,
+                           drake::systems::Context<double>& context_wo_spr) = 0;
+  // Finalize and ensure that users construct OscTrackingDataMBP derived class
+  // correctly.
+  virtual void CheckDerivedOscTrackingData() = 0;
+
+  // Trajectory name
+  std::string name_;
+
+  // Dimension of the traj
+  int n_r_;
+
+  // PD control gains
+  Eigen::MatrixXd K_p_;
+  Eigen::MatrixXd K_d_;
+
+  // Cost weights
+  Eigen::MatrixXd W_;
+
+  // cache
```

Comment:
This comment should probably be removed (doesn't fit here)

Reply:
Done.